django/thunderstore/repository/api/experimental/serializers.py

```diff
@@ -95,3 +98,104 @@ class Meta:
             "has_nsfw_content",
             "categories",
         )
+
+
+class PackageUploadAuthorNameField(serializers.SlugRelatedField):
+    """Package upload's author name metadata field."""
+
+    def __init__(self, *args, **kwargs):
+        kwargs["slug_field"] = "name"
+        super().__init__(*args, **kwargs)
+
+    def get_queryset(self):
+        return UploaderIdentity.objects.filter(
+            members__user=self.context["request"].user,
+        )
+
+
+class PackageUploadCategoriesField(serializers.RelatedField):
+    """Package upload's categories metadata field."""
+
+    def get_queryset(self):
+        return PackageCategory.objects.filter(
+            community=self.context["request"].community,
+        )
+
+    def to_representation(self, value):
+        return [c.slug for c in value]
+
+    def to_internal_value(self, data):
+        if not isinstance(data, list):
+            raise serializers.ValidationError("Not a list")
+
+        out = []
+        for category_slug in data:
+            category = self.get_queryset().filter(slug=category_slug).first()
+            if not category:
+                raise serializers.ValidationError(
+                    f"'{category_slug}' category not found",
+                )
+            out.append(category)
+        return out
+
+
+class DictSerializer(serializers.Serializer):
+    def create(self, validated_data):
+        return validated_data
+
+
+class JSONToSerializerField(serializers.Field):
+    """Parses a JSON string and passes the data to a DictSerializer subclass."""
+
+    def __init__(self, serializer, *args, **kwargs):
+        self._serializer = serializer
+        super().__init__(*args, **kwargs)
+
+    def to_internal_value(self, data):
+        try:
+            serializer = self._serializer(data=json.loads(data), context=self.context)
+        except json.JSONDecodeError:
+            raise serializers.ValidationError("Invalid JSON")
+        serializer.is_valid(raise_exception=True)
+        return serializer.save()
+
+    def to_representation(self, value):
+        raise NotImplementedError()
+
+
+class PackageUploadMetadataSerializer(DictSerializer):
+    """Non-file fields used for package upload."""
+
+    author_name = PackageUploadAuthorNameField()
+    categories = PackageUploadCategoriesField()
+    has_nsfw_content = serializers.BooleanField()
+
+
+class PackageUploadSerializerExperiemental(serializers.Serializer):
+    file = serializers.FileField(write_only=True)
+    metadata = JSONToSerializerField(serializer=PackageUploadMetadataSerializer)
+
+    def _create_form(self, data) -> PackageUploadForm:
+        request = self.context["request"]
+        metadata = data.get("metadata", {})
+        return PackageUploadForm(
+            request.user,
+            metadata.get("author_name"),
```

Comment:
I'd use keywoard arguments rather than positional just for clarity. Makes it easier for reviewing at the very least